docs/proposals/pvcs-smaller-than-1gi.md

```diff
@@ -0,0 +1,706 @@
+
+```protobuf
+    // Represents the input for CreateLV.
+    message CreateLVRequest {
+      string name = 1;              // The logical volume name.
++/-   uint64 size_gb = 2 [deprecated = true];  // Volume size in GiB. Deprecated in favor of size
+      repeated string tags = 3;     // Tags to add to the volume during creation
+      string device_class = 4;
+      string lvcreate_option_class = 5;
+  
++     int64 size = 6; // Volume size in canonical bytes.
+    }
+```
+
+Note that SizeGB is used as uint64 where we cast down to int64. This means that potentially,
+if someone had a volume before this change greater than `9.223.372.036.854.775.807 Gi`, he would now experience an overflow
+that would break lvmd. *However*, we need to be aware that the CSIDriver capacity-ranges at most support values up to int64 limits,
+so we do not break any currently known path.
+
+This change from `size_gb` to `size` can be reused across all Requests/Responses and only needs minor adjustment.
+
+There is a second part to this change: the json parsing in `lvmd/command`: This is currently the root of why all comparisons need to be done with `uint64` instead of `int64`.
+
+We can replace all occurrences during the parsing process to get arround this:
+
+```go
+    type vg struct {
+        name string
+        uuid string
++/-     size int64
++/-     free int64
+    }
+    
+    func (u *vg) UnmarshalJSON(data []byte) error {
+        type vgInternal struct {
+            Name string `json:"vg_name"`
+            UUID string `json:"vg_uuid"`
+            Size string `json:"vg_size"`
+            Free string `json:"vg_free"`
+        }
+    
+        var temp vgInternal
+        if err := json.Unmarshal(data, &temp); err != nil {
+            return err
+        }
+    
+        u.name = temp.Name
+        u.uuid = temp.UUID
+    
+        var convErr error
++/-     u.size, convErr = strconv.ParseInt(temp.Size, 10, 64)
+        if convErr != nil {
+            return convErr
+        }
++/-     u.free, convErr = strconv.ParseInt(temp.Free, 10, 64)
+        if convErr != nil {
+            return convErr
+        }
+    
+        return nil
+    }
+```
+
+Of course the same adjustments from `uint64` to `int64` be done for all methods using these structs and similar structs like `lv`.  
+Then they can be reused in `lvmd/lvservice` and `lvmd/vgservice` to simplify the calls from uint64 bitshift comparisons to simple byte comparisons:
+The resulting `CreateLV` call for example would look almost exactly like the original `CreateLV`:
+
+```go
+    func (s *lvService) CreateLV(_ context.Context, req *proto.CreateLVRequest) (*proto.CreateLVResponse, error) {
+        dc, err := s.dcmapper.DeviceClass(req.DeviceClass)
+        // ...
+        vg, err := command.FindVolumeGroup(dc.VolumeGroup)
+        // ...
+        oc := s.ocmapper.LvcreateOptionClass(req.LvcreateOptionClass)
++/-     requested := req.GetSize()
```

Comment:
When updating, the old lvmd and the new lvmd work simultaneously. Also, updating the controller and lvmd is not atomic.
Should we use `size_bg` when `size` is missing? Also, should we set `size_gb` along with `size` in response?

Reply:
That is actually a great idea and concern! Like you suggested, In this case we should probably write a logic that is able to work with both like you suggested:

Assuming 
```protobuf
    // Represents the input for CreateLV.
    message CreateLVRequest {
      string name = 1;              // The logical volume name.
+/-   uint64 size_gb = 2 [deprecated = true];
      repeated string tags = 3;     // Tags to add to the volume during creation
      string device_class = 4;
      string lvcreate_option_class = 5;
+     int64 size = 6; // Volume size in bytes
    }
```

we can write 

```go
if req.GetSize() != 0 {
   requested = req.GetSize()
} else {
   requested := int64(req.GetSizeGb() << 30)
}
```

and just set both in the request like you suggested.

That saves us in the asynchronous updates of lvmd and topolvm because both will be able to deal with this. However I want to add that technically, after redeployment is completed, the errors would disappear so this might be acceptable.

Reply:
I think my suggestion is safe, but it is a rare case where there are in-flight requests when updating, and it does not affect the high-level design, so you are free to do this or not when implementing.